lib/evmone/eof.cpp

```diff
@@ -718,18 +749,35 @@ EOFValidationError validate_eof1(evmc_revision rev, bytes_view main_container) n
         if (referenced_by_eofcreate && !header.can_init(container.size()))
             return EOFValidationError::eofcreate_with_truncated_container;
 
+        // Check if container kind (which instructions it has inside) is compatible with
+        // instructions that reference it in parent container.
+        if (container_kind == ContainerKind::initcode_container)
+        {
+            if (referenced_by_returncontract)
+                return EOFValidationError::incompatible_container_kind;
+        }
+        else if (container_kind == ContainerKind::runtime_container)
+        {
+            if (referenced_by_eofcreate)
+                return EOFValidationError::incompatible_container_kind;
+        }
+
```

Comment:
Validation whether subcontainer is referenced by anything can be added here, or below in "enqueue subcontainers" loop, better in a separate PR.

Reply:
added